File: 12.py
from __future__ import annotations

import heapq
import time
from libs import *
from typing import Generator


def get_neighbors(x: int, y: int) -> Generator[tuple[int, int], None, None]:
    yield x - 1, y
    yield x, y - 1
    yield x + 1, y
    yield x, y + 1


def bfs(
    coords: dict[tuple[int, int], str],
    start: tuple[int, int],
    end: tuple[int, int],
    part=1,
) -> int:
    visited: list[tuple[int, int]] = [end]
    queue: list[tuple[int, tuple[int, int]]] = []
    heapq.heappush(queue, (0, end))

    while queue:
        dist, last_coord = heapq.heappop(queue)
        x, y = last_coord
        val = coords[last_coord]

        for i, j in get_neighbors(x, y):
            if (i, j) not in coords:
                continue
            if (i, j) in visited:
                continue
            if ord(coords[(i, j)]) - ord(val) < -1:
                continue

            if part == 1:
                if (i, j) == start:
                    return dist + 1
            else:
                if coords[(i, j)] == "a":
                    return dist + 1

            visited.append((i, j))
            heapq.heappush(queue, (dist + 1, (i, j)))

    return 0


def main():
    print("--- Running Day 12 ---")

    start = time.perf_counter()

    try:
        with open("12.txt", "r") as f:
            data = f.readlines()
    except Exception as e:
        raise ValueError("Unable to read file") from e

    coords: dict[tuple[int, int], str] = {}
    start_point = (0, 0)
    end_point = (-1, -1)
    for row, line in enumerate(data):
        for col, chr in enumerate(line):
            if chr == "S":
                start_point = (row, col)
                coords[(row, col)] = "a"
            elif chr == "E":
                end_point = (row, col)
                coords[(row, col)] = "z"
            else:
                coords[(row, col)] = chr

    print(f"Output Part 1: {bfs(coords, start_point, end_point)}")
    print(f"Output Part 2: {bfs(coords, start_point, end_point, 2)}")

    end = time.perf_counter()
    rtime = (end - start) * 1000  # sec -> ms
    print(f"Took {format_runtime(rtime)}\n")


if __name__ == "__main__":
    main()

File: 12.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const { formatRuntime } = require('./libs');

class Queue {
  constructor() {
    this.elements = {};
    this.head = 0;
    this.tail = 0;
  }
  enqueue(element) {
    this.elements[this.tail] = element;
    this.tail++;
  }
  dequeue() {
    const item = this.elements[this.head];
    delete this.elements[this.head];
    this.head++;
    return item;
  }
  peek() {
    return this.elements[this.head];
  }
  get length() {
    return this.tail - this.head;
  }
  get isEmpty() {
    return this.length === 0;
  }
}

async function readInput() {
  const input = await fs.promises.readFile('12.txt', 'utf-8');
  return input.trim().split('\n');
}

const getNeighbours = (x, y) => {
  const neighbours = [];
  neighbours.push({ x: x - 1, y: y });
  neighbours.push({ x: x, y: y - 1 });
  neighbours.push({ x: x + 1, y: y });
  neighbours.push({ x: x, y: y + 1 });
  return neighbours;
};

const bfs = (coords, start, end, part) => {
  const visited = new Set();
  visited.add(JSON.stringify(end));
  const queue = new Queue();
  queue.enqueue({ dist: 0, coord: end });

  while (!queue.isEmpty) {
    const item = queue.dequeue();
    const dist = item.dist;
    const x = item.coord.x;
    const y = item.coord.y;

    const val = coords.get(JSON.stringify({ x: x, y: y }));

    const neighbours = getNeighbours(x, y);
    for (let i = 0; i < neighbours.length; i++) {
      const coord = neighbours[i];
      if (coords.get(JSON.stringify(coord)) === undefined) {
        continue;
      }
      if (visited.has(JSON.stringify(coord))) {
        continue;
      }
      if (
        coords.get(JSON.stringify(coord)).charCodeAt() - val.charCodeAt() <
        -1
      ) {
        continue;
      }
      if (part === 1) {
        if (coord.x === start.x && coord.y === start.y) {
          return dist + 1;
        }
      } else {
        if (coords.get(JSON.stringify(coord)) == 'a') {
          return dist + 1;
        }
      }
      visited.add(JSON.stringify(coord));
      queue.enqueue({ dist: dist + 1, coord: coord });
    }
  }
  return 0;
};

async function puzzle() {
  console.log('--- Running Day 12 ---');

  const start = performance.now();

  const input = await readInput();

  const coords = new Map();
  let startPoint = { x: 0, y: 0 };
  let endPoint = { x: -1, y: -1 };

  for (let row = 0; row < input.length; row++) {
    const line = input[row];
    for (let col = 0; col < line.length; col++) {
      const chr = line[col];
      if (chr === 'S') {
        startPoint = { x: row, y: col };
        coords.set(JSON.stringify(startPoint), 'a');
      } else if (chr === 'E') {
        endPoint = { x: row, y: col };
        coords.set(JSON.stringify(endPoint), 'z');
      } else {
        coords.set(JSON.stringify({ x: row, y: col }), chr);
      }
    }
  }

  console.log('Part 1: ' + bfs(coords, startPoint, endPoint, 1));
  console.log('Part 2: ' + bfs(coords, startPoint, endPoint, 2));

  const end = performance.now();
  const rtime = end - start;
  console.log('Took:', formatRuntime(rtime));
}

puzzle();
